cmd/bedSimilarity/bedSimilarity.go

```diff
@@ -0,0 +1,202 @@
+package main
+
+//TODO : add check for self-overlap bed files
+//TODO: add enrichment calculations
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/bed"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/interval"
+	"log"
+	"strings"
+)
+
+type settings struct {
+	bedA             string
+	bedB             string
+	outFile          string
+	list             string
+	matrixAverage    string
+	matrixComponents string
+}
+
+type matrixLine struct {
+	name string
+	vals []float64
+}
+
+func compareTwo(s settings) {
+	var out []string
+	a := bed.Read(s.bedA)
+	b := bed.Read(s.bedB)
+	aName := separatePath(s.bedA)
+	bName := separatePath(s.bedB)
+	intervalsA := interval.BedSliceToIntervals(a)
+	intervalsB := interval.BedSliceToIntervals(b)
+	overlapsA, overlapsB, overlapAverage := interval.IntervalSimilarity(intervalsA, intervalsB)
+	header := fmt.Sprintf("proportion overlaps of %s in %s\tproportion overlaps of %s in %s\tbedSimilarityScore", aName, bName, bName, aName)
+	data := fmt.Sprintf("%f\t%f\t%f", overlapsA, overlapsB, overlapAverage)
+	out = append(out, header)
+	out = append(out, data)
+	fileio.Write(s.outFile, out)
+}
+
+func multipleComparisons(s settings) {
+	var a, b []bed.Bed
+	var intervalsA, intervalsB []interval.Interval
+	var percA, percB, avg float64
+	var outMatrix *fileio.EasyWriter
+	var err error
+	var j int
+	var l matrixLine
+	var aName, bName string
+	var allFiles []string = []string{"x"}
+
+	out := fileio.EasyCreate(s.outFile)
+	fileio.WriteToFileHandle(out, "A\tB\tproportion overlaps of A in B\tproportion overlaps of B in A\tbedSimilarityScore")
+
+	if s.matrixAverage != "" {
+		outMatrix = fileio.EasyCreate(s.matrixAverage)
+	}
+	if s.matrixComponents != "" {
+		outMatrix = fileio.EasyCreate(s.matrixComponents)
+	}
+
+	files := fileio.Read(s.list)
+
+	if s.matrixAverage != "" || s.matrixComponents != "" {
+		for _, i := range files {
+			allFiles = append(allFiles, separatePath(i))
+		}
+		header := strings.Join(allFiles, "\t")
+		fileio.WriteToFileHandle(outMatrix, header)
+	}
+
+	for i := range files {
```

Comment:
If I am following your variable names correctly, something like this should work:

```golang
func multipleComparisons(s settings) {
	// var a, b []bed.Bed
	var intervalsA, intervalsB []interval.Interval
	var percA, percB, avg float64
	var outMatrix *fileio.EasyWriter
	var err error
	var j int
	var l matrixLine
	var aName, bName string
	var allFiles []string = []string{"x"}

	out := fileio.EasyCreate(s.outFile)
	fileio.WriteToFileHandle(out, "A\tB\tproportion overlaps of A in B\tproportion overlaps of B in A\tbedSimilarityScore")

	if s.matrixAverage != "" {
		outMatrix = fileio.EasyCreate(s.matrixAverage)
	}
	if s.matrixComponents != "" {
		outMatrix = fileio.EasyCreate(s.matrixComponents)
	}

	files := fileio.Read(s.list)

	if s.matrixAverage != "" || s.matrixComponents != "" {
		for _, i := range files {
			allFiles = append(allFiles, separatePath(i))
		}
		header := strings.Join(allFiles, "\t")
		fileio.WriteToFileHandle(outMatrix, header)
	}

	// Read data from files once and store
	numFiles := len(files)
	fileData := make([][]bed.Bed, numFiles)
	intervalData := make([][]interval.Interval, numFiles)
	names := make([]string, numFiles)

	for i, file := range files {
		fileData[i] = bed.Read(file)
		intervalData[i] = interval.BedSliceToIntervals(fileData[i])
		names[i] = separatePath(file)
	}

	for i := 0; i < numFiles; i++ {
		l.name = names[i]
		aName = names[i]
		l.vals = []float64{}
		// a = fileData[i]
		intervalsA = intervalData[i]
		for j = range files {
			if files[i] == files[j] {
				if s.matrixComponents != "" || s.matrixAverage != "" {
					l.vals = append(l.vals, 1.0)
				}
				continue
			}
			// b = fileData[j]
			bName = names[j]
			intervalsB = intervalData[j]
			percA, percB, avg = interval.IntervalSimilarity(intervalsA, intervalsB)
			if j > i {
				fileio.WriteToFileHandle(out, fmt.Sprintf("%s\t%s\t%f\t%f\t%f", aName, bName, percA, percB, avg))
			}

			if s.matrixAverage != "" {
				l.vals = append(l.vals, avg)
			} else if s.matrixComponents != "" {
				l.vals = append(l.vals, percA)
			}
		}
		if s.matrixAverage != "" || s.matrixComponents != "" {
			writeMatrixLine(l, outMatrix)
		}
	}

	err = out.Close()
	exception.PanicOnErr(err)
	if s.matrixAverage != "" || s.matrixComponents != "" {
		err = outMatrix.Close()
		exception.PanicOnErr(err)
	}
}
```
- I added the loop to read and process the files and storing the data/names/etc. and commented out `fileStorage` so you can easily follow where I made modifications. This should be a working implementation (i think)

Reply:
this looks awesome, I'll dive into this deeper today!

Reply:
Made these improvements thanks! The only think i changed from your suggested code is that instead of having the intermediate [][]bed.Bed to create the [][]interval.Interval, I put the bed.Read(file) right into the bedSliceToIntervals function, which eliminated the need for the fileData variable